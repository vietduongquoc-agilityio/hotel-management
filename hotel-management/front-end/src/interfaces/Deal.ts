export interface DealData {
  dealName: string;
  referenceNumber: string;
  startDate: Date;
  endDate: Date;
  roomType: string;
  statusDeal: string;
  reservationsLeft: number;
  documentId: string;
}

export type NewDealData = Omit<DealData, "documentId">;
