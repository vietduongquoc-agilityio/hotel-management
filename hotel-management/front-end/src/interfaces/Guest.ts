export interface GuestData {
    guestName: string;
    roomType: string;
    stay: number;
    price: number;
    registrationNumber: string;
    totalAmount: number;
    checkInDate: Date;
    documentId: string;
}

export type NewGuestData = Omit<GuestData, "documentId">;
