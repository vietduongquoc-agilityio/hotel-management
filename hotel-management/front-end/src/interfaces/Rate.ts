export interface RateData {
  roomType: string;
  cancellationPolicy: string;
  deals: string;
  dealPrice: string;
  availability: string;
  documentId: string;
  rate: string;
  totalOfBooked?: number;
}

export type NewRateData = Omit<RateData, "documentId">;
