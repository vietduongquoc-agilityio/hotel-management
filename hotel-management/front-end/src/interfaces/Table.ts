export interface EditFunctionType<T> {
  initialData: T;
  onEditFunction: (updatedData: T) => void;
}

export interface DeleteFunctionType {
  documentId: string;
  onDeleteFunction: (id: string) => void;
}
