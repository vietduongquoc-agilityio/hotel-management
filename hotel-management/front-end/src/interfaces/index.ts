export * from "./Rate";
export * from "./Room";
export * from "./Table";
export * from "./Guest"
