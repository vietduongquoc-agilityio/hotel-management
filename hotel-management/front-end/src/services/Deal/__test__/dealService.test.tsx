import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import {
  getDeals,
  createDealApi,
  updateDeal,
  deleteDeal,
} from "@/services";
import { NewDealData } from "@/interfaces";

jest.mock("@chakra-ui/react", () => ({
  createStandaloneToast: jest.fn(() => ({
    toast: jest.fn(),
  })),
}));

const mock = new MockAdapter(axios);
const BASE_URL = process.env.VITE_BASE_URL;

describe("dealService", () => {
  afterEach(() => {
    mock.reset();
    jest.clearAllMocks();
  });

  describe("getDeals", () => {
    it("should fetch deals with pagination and match snapshot", async () => {
      const mockResponse = {
        data: [
          { id: "1", attributes: { dealName: "Deal 1", roomType: "Deluxe" } },
          { id: "2", attributes: { dealName: "Deal 2", roomType: "Standard" } },
        ],
        meta: { pagination: { page: 1, pageSize: 10, pageCount: 1, total: 2 } },
      };

      mock.onGet(`${BASE_URL}/deals`).reply(200, mockResponse);

      const result = await getDeals(1, 10);
      expect(result).toMatchSnapshot();
    });

    it("should return error message on failure", async () => {
      mock.onGet(`${BASE_URL}/deals`).reply(500);

      const result = await getDeals(1, 10);
      expect(result).toEqual({ message: "Error fetching deal data", data: null });
    });
  });

  describe("createDealApi", () => {
    it("should create a new deal and match snapshot", async () => {
      const dealData: NewDealData = {
        dealName: "Special Offer",
        referenceNumber: "12345",
        startDate: new Date(),
        endDate: new Date(),
        roomType: "Deluxe",
        statusDeal: "Active",
        reservationsLeft: 5,
      };
      const mockResponse = { data: { id: "1", attributes: { ...dealData } } };

      mock.onPost(`${BASE_URL}/deals`).reply(200, mockResponse);

      const result = await createDealApi(dealData);
      expect(result).toMatchSnapshot();
    });

    it("should return error message on failure", async () => {
      const dealData: NewDealData = {
        dealName: "Special Offer",
        referenceNumber: "12345",
        startDate: new Date(),
        endDate: new Date(),
        roomType: "Deluxe",
        statusDeal: "Active",
        reservationsLeft: 5,
      };

      mock.onPost(`${BASE_URL}/deals`).reply(500);

      const result = await createDealApi(dealData);
      expect(result).toEqual({ message: "Error updated deal details", data: null });
    });
  });

  describe("updateDeal", () => {
    it("should update a deal by ID and match snapshot", async () => {
      const dealId = "1";
      const dealData: NewDealData = {
        dealName: "Updated Deal",
        referenceNumber: "54321",
        startDate: new Date(),
        endDate: new Date(),
        roomType: "Suite",
        statusDeal: "Inactive",
        reservationsLeft: 3,
      };
      const mockResponse = {
        data: { id: dealId, attributes: { ...dealData } },
      };

      mock.onPut(`${BASE_URL}/deals/${dealId}`).reply(200, mockResponse);

      const result = await updateDeal(dealId, dealData);
      expect(result).toMatchSnapshot();
    });

    it("should throw error if dealId is missing", async () => {
      const dealData: NewDealData = {
        dealName: "Updated Deal",
        referenceNumber: "54321",
        startDate: new Date(),
        endDate: new Date(),
        roomType: "Suite",
        statusDeal: "Inactive",
        reservationsLeft: 3,
      };

      await expect(updateDeal("", dealData)).rejects.toThrow(
        "Missing document ID for deal update."
      );
    });

    it("should return error message on failure", async () => {
      const dealId = "1";
      const dealData: NewDealData = {
        dealName: "Updated Deal",
        referenceNumber: "54321",
        startDate: new Date(),
        endDate: new Date(),
        roomType: "Suite",
        statusDeal: "Inactive",
        reservationsLeft: 3,
      };

      mock.onPut(`${BASE_URL}/deals/${dealId}`).reply(500);

      const result = await updateDeal(dealId, dealData);
      expect(result).toEqual({ message: "Error updated deal details", data: null });
    });
  });

  describe("deleteDeal", () => {
    it("should delete a deal by ID and match snapshot", async () => {
      const dealId = "1";
      const mockResponse = { message: "Deal deleted successfully" };

      mock.onDelete(`${BASE_URL}/deals/${dealId}`).reply(200, mockResponse);

      const result = await deleteDeal(dealId);
      expect(result).toMatchSnapshot();
    });

    it("should return error message on failure", async () => {
      const dealId = "1";
      mock.onDelete(`${BASE_URL}/deals/${dealId}`).reply(500);

      const result = await deleteDeal(dealId);
      expect(result).toEqual({ message: "Error deleted deal details", data: null });
    });
  });
});
