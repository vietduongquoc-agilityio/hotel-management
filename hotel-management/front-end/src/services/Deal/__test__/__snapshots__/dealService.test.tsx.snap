// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dealService createDealApi should create a new deal and match snapshot 1`] = `
{
  "data": {
    "attributes": {
      "dealName": "Special Offer",
      "endDate": "2024-12-25T01:48:58.994Z",
      "referenceNumber": "12345",
      "reservationsLeft": 5,
      "roomType": "Deluxe",
      "startDate": "2024-12-25T01:48:58.994Z",
      "statusDeal": "Active",
    },
    "id": "1",
  },
}
`;

exports[`dealService deleteDeal should delete a deal by ID and match snapshot 1`] = `
{
  "data": {
    "message": "Deal deleted successfully",
  },
  "message": "Deal deleted successfully",
}
`;

exports[`dealService getDeals should fetch deals with pagination and match snapshot 1`] = `
{
  "deals": [
    {
      "attributes": {
        "dealName": "Deal 1",
        "roomType": "Deluxe",
      },
      "id": "1",
    },
    {
      "attributes": {
        "dealName": "Deal 2",
        "roomType": "Standard",
      },
      "id": "2",
    },
  ],
  "pagination": {
    "page": 1,
    "pageCount": 1,
    "pageSize": 10,
    "total": 2,
  },
}
`;

exports[`dealService updateDeal should update a deal by ID and match snapshot 1`] = `
{
  "data": {
    "data": {
      "attributes": {
        "dealName": "Updated Deal",
        "endDate": "2024-12-25T01:48:59.000Z",
        "referenceNumber": "54321",
        "reservationsLeft": 3,
        "roomType": "Suite",
        "startDate": "2024-12-25T01:48:59.000Z",
        "statusDeal": "Inactive",
      },
      "id": "1",
    },
  },
  "message": "Deal updated successfully",
}
`;
